### PR TITLE
Introduce ProviderMarketStrategy

### PR DESCRIPTION
**Pull request by mplebanski**
Introduce ProviderMarketStrategy (and its subclasses) to allow provider to have specific market strategy for every task type.

This PR is part of the usage market epic.

### CODE REVIEW

**Review by mbenke — APPROVED** [review]
LGTM

File: apps/wasm/task.py [hunk 1]
```diff
@@ -28,7 +29,8 @@
 from apps.core.task.coretaskstate import Options, TaskDefinition
 from apps.wasm.environment import WasmTaskEnvironment
 from golem.marketplace.wasm_marketplace import (
-    RequestorWasmMarketStrategy, UsageReport
+    RequestorWasmMarketStrategy, UsageReport,
```
Comment (maaktweluit) [inline apps/wasm/task.py]: one import per line according to the style guide ( when they do not fit on a single one )

File: golem/marketplace/__init__.py [hunk 2]
```diff
@@ -1,7 +1,18 @@
 # pylint: disable=unused-import
 from .marketplace import Offer, ProviderPerformance  # noqa
-from .marketplace import RequestorMarketStrategy  # noqa
-from .brass_marketplace import RequestorBrassMarketStrategy  # noqa
-from .wasm_marketplace import RequestorWasmMarketStrategy  # noqa
+from .marketplace import (  # noqa
```
Comment (maaktweluit) [inline golem/marketplace/__init__.py:2]: you can add `Offer` and `ProviderPerformance` to this block

File: golem/marketplace/brass_marketplace.py [hunk 3]
```diff
@@ -52,8 +53,34 @@ def resolve_task_offers(cls, task_id: str) -> Optional[List[Offer]]:
...
         def payment_computer(price: int):
             return price * task.header.subtask_timeout
+
         return payment_computer
+
+
+class ProviderBrassMarketStrategy(ProviderMarketStrategy):
+
+    @classmethod
+    def calculate_price(cls, pricing: ProviderPricing, max_price: int,
```
Comment (maaktweluit) [inline golem/marketplace/brass_marketplace.py]: linebreaks

File: golem/marketplace/wasm_marketplace.py [hunk 4]
```diff
@@ -1,13 +1,17 @@
 import logging
-from typing import (
-    Callable, List, Dict, ClassVar, Tuple, Optional,
-    Iterable, TYPE_CHECKING
-)
+from typing import (Callable, List, Dict, ClassVar, Tuple, Optional, Iterable,
```
Comment (maaktweluit) [inline golem/marketplace/wasm_marketplace.py:2]: this is OK, typing is exempt from the one var per line rule ( just FYI )

File: tests/golem/marketplace/test_marketplace.py [hunk 5]
```diff
@@ -104,12 +103,20 @@ class TestRequestorBrassMarketStrategy(TestCase):
 
     @staticmethod
     def _mock_offer():
-        return Offer(
-            provider_id='provider_1',
-            provider_performance=ProviderPerformance(100),
-            max_price=5000,
-            price=5000
-        )
+        mock_offer = MagicMock(spec_set=[
```
Comment (maaktweluit) [inline tests/golem/marketplace/test_marketplace.py:106]: can we use `spec=TYPE` here?

**Review by maaktweluit — APPROVED** [review]
LGTM!

Few style comments ( that we should start catching with the linter ).
Logic seems good, so approving :)

File: apps/wasm/task.py (hunk 1 above)
Comment (mbenke) [reply apps/wasm/task.py]: done

File: golem/marketplace/__init__.py (hunk 2 above)
Comment (mbenke) [reply golem/marketplace/__init__.py:2]: done

File: golem/marketplace/brass_marketplace.py (hunk 3 above)
Comment (mbenke) [reply golem/marketplace/brass_marketplace.py]: done

File: tests/golem/marketplace/test_marketplace.py (hunk 5 above)
Comment (mbenke) [reply tests/golem/marketplace/test_marketplace.py:106]: It doesn't work  for me, but I am no pytest expert. Leaving it as is, if you find a way to make it work, be my guest

**Pull request merged**